DragaliaAPI/Features/Missions/MissionProgressionService.cs

```diff
@@ -87,55 +245,75 @@ public async Task ProcessMissionEvents()
+            this.eventQueue.ToList()
+        );
 
-            IEnumerable<int> affectedMissions = info.Requirements
-                .Where(
-                    x =>
-                        (x.Parameter is null || x.Parameter == evt.Parameter)
-                        && (x.Parameter2 is null || x.Parameter2 == evt.Parameter2)
-                        && (x.Parameter3 is null || x.Parameter3 == evt.Parameter3)
-                        && (x.Parameter4 is null || x.Parameter4 == evt.Parameter4)
-                )
-                .SelectMany(x => x.Missions)
-                .Select(x => x.Id)
-                .ToList();
+        while (this.eventQueue.TryDequeue(out MissionEvent? evt))
+        {
+            List<(MissionType Type, int Id)> affectedMissions =
+                MasterAsset.MissionProgressionInfo.Enumerable
+                    .Where(x => x.CompleteType == evt.Type)
+                    .Where(
+                        x =>
+                            (x.Parameter is null || x.Parameter == evt.Parameter)
+                            && (x.Parameter2 is null || x.Parameter2 == evt.Parameter2)
+                            && (x.Parameter3 is null || x.Parameter3 == evt.Parameter3)
+                            && (x.Parameter4 is null || x.Parameter4 == evt.Parameter4)
+                    )
+                    .Select(x => (x.MissionType, x.MissionId))
+                    .ToList();
 
             if (affectedMissions.Any())
             {
-                missionList ??= await this.missionRepository.Missions
+                missionList ??= await missionRepository.Missions
                     .Where(x => x.State == MissionState.InProgress)
                     .ToListAsync();
 
                 foreach (
                     DbPlayerMission progressingMission in missionList.Where(
-                        x => affectedMissions.Contains(x.Id) && x.State == MissionState.InProgress
+                        x =>
+                            affectedMissions.Contains((x.Type, x.Id))
+                            && x.State == MissionState.InProgress
                     )
                 )
                 {
                     Mission mission = Mission.From(progressingMission.Type, progressingMission.Id);
-                    progressingMission.Progress++;
+
+                    MissionProgressionInfo progressionInfo =
+                        MasterAsset.MissionProgressionInfo.Enumerable.Single(
+                            x =>
+                                x.MissionType == progressingMission.Type
+                                && x.MissionId == progressingMission.Id
+                        );
+
+                    if (progressionInfo.UseTotalValue)
+                    {
+                        if (progressingMission.Progress >= evt.TotalValue)
+                            continue;
+
+                        progressingMission.Progress = evt.TotalValue;
```

Comment:
what happens if you overshoot a total requirement? will the mission progress ever get set to mission.completevalue?

Reply:
true, forgot about that case